ts/src/woofipro.ts

```diff
@@ -0,0 +1,2636 @@
+        if (isMarket && !isStop) {
+            request[orderQtyKey] = this.amountToPrecision (symbol, amount);
+        } else if (algoType !== 'POSITIONAL_TP_SL') {
+            request[orderQtyKey] = this.amountToPrecision (symbol, amount);
+        }
+        const clientOrderId = this.safeStringN (params, [ 'clOrdID', 'clientOrderId', 'client_order_id' ]);
+        if (clientOrderId !== undefined) {
+            request['client_order_id'] = clientOrderId;
+        }
+        if (stopPrice !== undefined) {
+            request['trigger_price'] = this.priceToPrecision (symbol, stopPrice);
+            request['algo_type'] = 'STOP';
+        } else if ((stopLoss !== undefined) || (takeProfit !== undefined)) {
+            request['algo_type'] = 'TP_SL';
+            const outterOrder = {
+                'symbol': market['id'],
+                'reduce_only': false,
+                'algo_type': 'POSITIONAL_TP_SL',
+                'child_orders': [],
+            };
+            const closeSide = (orderSide === 'BUY') ? 'SELL' : 'BUY';
+            if (stopLoss !== undefined) {
+                const stopLossPrice = this.safeNumber2 (stopLoss, 'triggerPrice', 'price', stopLoss);
+                const stopLossOrder = {
+                    'side': closeSide,
+                    'algo_type': 'TP_SL',
+                    'trigger_price': this.priceToPrecision (symbol, stopLossPrice),
+                    'type': 'LIMIT',
+                    'reduce_only': true,
+                };
+                outterOrder['child_orders'].push (stopLossOrder);
+            }
+            if (takeProfit !== undefined) {
+                const takeProfitPrice = this.safeNumber2 (takeProfit, 'triggerPrice', 'price', takeProfit);
+                const takeProfitOrder = {
+                    'side': closeSide,
+                    'algo_type': 'TP_SL',
+                    'trigger_price': this.priceToPrecision (symbol, takeProfitPrice),
+                    'type': 'LIMIT',
+                    'reduce_only': true,
+                };
+                outterOrder['child_orders'].push (takeProfitOrder);
+            }
+            request['child_orders'] = [ outterOrder ];
+        }
+        params = this.omit (params, [ 'clOrdID', 'clientOrderId', 'client_order_id', 'postOnly', 'timeInForce', 'stopPrice', 'triggerPrice', 'stopLoss', 'takeProfit', 'trailingPercent', 'trailingAmount', 'trailingTriggerPrice' ]);
+        let response = undefined;
+        if (isStop) {
+            response = await this.v1PrivatePostAlgoOrder (this.extend (request, params));
+            //
+            // {
+            //     "success": true,
+            //     "timestamp": 1702989203989,
+            //     "data": {
+            //       "order_id": 13,
+            //       "client_order_id": "testclientid",
+            //       "algo_type": "STOP",
+            //       "quantity": 100.12
+            //     }
+            // }
+            //
+        } else {
```

Comment:
also, is this error expected?
```
n woofipro createOrder "WIF/USDC:USDC" limit buy 1 2.5
Debugger attached.
2024-05-03T14:02:58.768Z
Node.js: v18.18.0
CCXT v4.3.14
woofipro.createOrder (WIF/USDC:USDC, limit, buy, 1, 2.5)
BadRequest woofipro {"success":false,"code":"-1005","message":"The order tag cannot be an empty string."}
```

Reply:
ok I added the tag but we need to double-check if it works well in the sandbox mode